File: internal/ops/loop_config.go
package ops

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type LoopConfig struct {
	MaxCallsPerHour           int
	CodexTimeoutMinutes       int
	MaxIterations             int
	MaxConsecutiveDoneSignals int
	MaxNoProgress             int
	MaxSameError              int
	MaxConsecutiveTestLoops   int
	MaxMissingStatus          int
}

func DefaultLoopConfig() LoopConfig {
	return LoopConfig{
		MaxCallsPerHour:           100,
		CodexTimeoutMinutes:       15,
		MaxIterations:             50,
		MaxConsecutiveDoneSignals: 2,
		MaxNoProgress:             3,
		MaxSameError:              5,
		MaxConsecutiveTestLoops:   3,
		MaxMissingStatus:          2,
	}
}

func LoadLoopConfig(path string) (LoopConfig, error) {
	cfg := DefaultLoopConfig()
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return cfg, nil
		}
		return cfg, err
	}
	scanner := bufio.NewScanner(strings.NewReader(string(data)))
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			return cfg, fmt.Errorf("invalid config line %d: %q (expected KEY=VALUE)", lineNo, line)
		}
		key := strings.TrimSpace(parts[0])
		val := strings.TrimSpace(parts[1])
		if hash := strings.Index(val, "#"); hash >= 0 {
			val = strings.TrimSpace(val[:hash])
		}
		if val == "" {
			return cfg, fmt.Errorf("empty value for %s on line %d", key, lineNo)
		}
		if err := applyLoopConfigValue(&cfg, key, val, lineNo); err != nil {
			return cfg, err
		}
	}
	if err := scanner.Err(); err != nil {
		return cfg, err
	}
	return cfg, nil
}

func applyLoopConfigValue(cfg *LoopConfig, key, val string, lineNo int) error {
	switch key {
	case "MAX_CALLS_PER_HOUR":
		return setLoopConfigInt(&cfg.MaxCallsPerHour, key, val, lineNo, true)
	case "CODEX_TIMEOUT_MINUTES":
		return setLoopConfigInt(&cfg.CodexTimeoutMinutes, key, val, lineNo, true)
	case "MAX_ITERATIONS":
		return setLoopConfigInt(&cfg.MaxIterations, key, val, lineNo, false)
	case "MAX_CONSECUTIVE_DONE_SIGNALS":
		return setLoopConfigInt(&cfg.MaxConsecutiveDoneSignals, key, val, lineNo, true)
	case "MAX_NO_PROGRESS":
		return setLoopConfigInt(&cfg.MaxNoProgress, key, val, lineNo, true)
	case "MAX_SAME_ERROR":
		return setLoopConfigInt(&cfg.MaxSameError, key, val, lineNo, true)
	case "MAX_CONSECUTIVE_TEST_LOOPS":
		return setLoopConfigInt(&cfg.MaxConsecutiveTestLoops, key, val, lineNo, true)
	case "MAX_MISSING_STATUS":
		return setLoopConfigInt(&cfg.MaxMissingStatus, key, val, lineNo, true)
	default:
		return fmt.Errorf("unknown config key %q on line %d", key, lineNo)
	}
}

func setLoopConfigInt(dst *int, key, val string, lineNo int, mustBePositive bool) error {
	parsed, err := strconv.Atoi(val)
	if err != nil {
		return fmt.Errorf("invalid int for %s on line %d: %q", key, lineNo, val)
	}
	if mustBePositive && parsed <= 0 {
		return fmt.Errorf("%s must be > 0 on line %d", key, lineNo)
	}
	if !mustBePositive && parsed < 0 {
		return fmt.Errorf("%s must be >= 0 on line %d", key, lineNo)
	}
	*dst = parsed
	return nil
}

File: src/ops/loop_config.rs
use std::fs;
use std::path::Path;

use crate::{LooprError, LooprResult};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LoopConfig {
    pub codex_timeout_minutes: i64,
    pub max_iterations: i64,
    pub max_missing_status: i64,
}

pub fn default_loop_config() -> LoopConfig {
    LoopConfig {
        codex_timeout_minutes: 15,
        max_iterations: 50,
        max_missing_status: 2,
    }
}

pub fn load_loop_config(path: &Path) -> LooprResult<LoopConfig> {
    let mut cfg = default_loop_config();
    let data = match fs::read_to_string(path) {
        Ok(value) => value,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(cfg),
        Err(err) => return Err(LooprError::new(format!("read {}: {}", path.display(), err))),
    };

    let mut line_no = 0;
    for line in data.lines() {
        line_no += 1;
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let mut parts = trimmed.splitn(2, '=');
        let key = match parts.next() {
            Some(value) => value.trim(),
            None => {
                return Err(LooprError::new(format!(
                    "invalid config line {}: \"{}\" (expected KEY=VALUE)",
                    line_no, trimmed
                )));
            }
        };
        let value = match parts.next() {
            Some(value) => value.trim(),
            None => {
                return Err(LooprError::new(format!(
                    "invalid config line {}: \"{}\" (expected KEY=VALUE)",
                    line_no, trimmed
                )));
            }
        };
        let mut val = value;
        if let Some(hash_idx) = val.find('#') {
            val = val[..hash_idx].trim();
        }
        if val.is_empty() {
            return Err(LooprError::new(format!(
                "empty value for {} on line {}",
                key, line_no
            )));
        }
        apply_loop_config_value(&mut cfg, key, val, line_no)?;
    }

    Ok(cfg)
}

fn apply_loop_config_value(
    cfg: &mut LoopConfig,
    key: &str,
    val: &str,
    line_no: usize,
) -> LooprResult<()> {
    match key {
        "CODEX_TIMEOUT_MINUTES" => {
            set_loop_config_int(&mut cfg.codex_timeout_minutes, key, val, line_no, true)
        }
        "MAX_ITERATIONS" => set_loop_config_int(&mut cfg.max_iterations, key, val, line_no, false),
        "MAX_MISSING_STATUS" => {
            set_loop_config_int(&mut cfg.max_missing_status, key, val, line_no, true)
        }
        _ => Ok(()),
    }
}

fn set_loop_config_int(
    dst: &mut i64,
    key: &str,
    val: &str,
    line_no: usize,
    must_be_positive: bool,
) -> LooprResult<()> {
    let parsed: i64 = val.parse().map_err(|_| {
        LooprError::new(format!(
            "invalid int for {} on line {}: \"{}\"",
            key, line_no, val
        ))
    })?;
    if must_be_positive && parsed <= 0 {
        return Err(LooprError::new(format!(
            "{} must be > 0 on line {}",
            key, line_no
        )));
    }
    if !must_be_positive && parsed < 0 {
        return Err(LooprError::new(format!(
            "{} must be >= 0 on line {}",
            key, line_no
        )));
    }
    *dst = parsed;
    Ok(())
}
